rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,116 @@
+
+### Teleport Auth server discovery and authentication
+
+To write something into Teleport, the manager needs to know what server to connect to and how. There're many options on how to discover Teleport instances in the cluster but for now, let's start with the simplest one - run manager as a sidecar container in the same pod as the Teleport auth server. Containers in the pod can share the volumes, so the manager container could have access to `/var/lib/teleport/proc` database, which contains administrator credentials. This is the same way how `tctl` utility works - it doesn't require any authentication when you run it on the same machine where Teleport is running.
+
+### High Availability mode
+
+One could deploy Teleport in High Availability mode, i.e. with `replicas` number greater than one and some highly available storage like DynamoDB or Firestore. When the Teleport pod is replicated, there're multiple sidecar containers as well. It's good to have a manager being highly available too, but we need to avoid a dispatch of the same event multiple times. To limit the concurrency a built-in Kubernetes leader election API should be used.
+
+### Resource definitions
+
+Custom resource definitions of Teleport resource look like this:
+
+```yaml
+apiVersion: resources.goteleport.com/v2
+kind: Role
+metadata:
+  name: folks
+  namespace: teleport
+spec:
+  allow:
+    request:
+      roles: ["admin"]
+---
+apiVersion: resources.goteleport.com/v4
+kind: User
+metadata:
+  name: human-bean
+  namespace: teleport
+spec:
+  roles: ["folks"]
+```
+
+Manager process watches for custom resource events. Once a custom resource is created/updated, the manager's watcher acknowledges it and upserts a resource into Teleport.
+
+The name of the Kubernetes custom resource is the name of the resource in Teleport. Also, every Kubernetes custom resource must have a `spec` field that has the same schema as a `spec` field of a resource in Teleport.
+
+Deletion of resources is being handled using [resource finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). Every acknowledged resource gets a finalizer named `"resources.goteleport.com/delete"` so when the deletion happens, Kubernetes doesn't actually delete it but sets a deletion timestamp. While the finalizer list is non-empty, the manager can still access the resource's fields. Deletion of the Kubernetes object leads to deletion of the associated resource in Teleport. Once a deletion in Teleport succeeds, the manager removes a finalizer from the Kubernetes object so it can be garbage-collected.
+
+### Identity definitions
+
+To run some plugin in the cluster, the credentials are required. Normally, one can get the credentials using `tctl auth sign` command or using `tsh login`. To automate the credentials generation, the `Identity` resource is introduced. The identity definition looks like this:
+
+```yaml
+apiVersion: control.goteleport.com/v8
```

Comment:
Should this be updated to `instances.goteleport.com/v8`? Seems like you changed it below.